qiskit/providers/ibmq/api_v2/clients/websocket.py

```diff
@@ -150,64 +153,81 @@ def get_job_status(self, job_id, timeout=None):
+                # a timeout has been reached.
+                while True:
+                    try:
+                        with warnings.catch_warnings():
+                            # Suppress websockets deprecation warnings until the fix is available
+                            warnings.filterwarnings("ignore", category=DeprecationWarning)
+                            if timeout:
+                                response_raw = yield from asyncio.wait_for(
+                                    websocket.recv(), timeout=timeout)
+
+                                # Decrease the timeout, with a 5-second grace period.
+                                elapsed_time = time.time() - start_time
+                                timeout = max(5, int(original_timeout - elapsed_time))
+                            else:
+                                response_raw = yield from websocket.recv()
+                        logger.debug('Received message from websocket: %s',
+                                     response_raw)
+
+                        response = WebsocketMessage.from_bytes(response_raw)
+                        last_status = response.data
+
+                        job_status = response.data.get('status')
+                        if (job_status and
+                                ApiJobStatus(job_status) in API_JOB_FINAL_STATES):
+                            break
+
+                    except futures.TimeoutError:
+                        # Timeout during our wait.
+                        raise WebsocketTimeoutError('Timeout reached') from None
+                    except ConnectionClosed as ex:
+                        # From the API:
+                        # 4001: closed due to an internal errors
+                        # 4002: closed on purpose (no more updates to send)
+                        # 4003: closed due to job not found.
+                        connection_closed_ex_code = ex.code
+                        message = 'Unexpected error'
+
+                        if ex.code == 4001:
+                            message = 'Internal server error'
+                        elif ex.code == 4002:
+                            break
+                        elif ex.code == 4003:
+                            message = 'Job id not found'
+                        raise WebsocketError('Connection with websocket closed '
+                                             'unexpectedly: {}'.format(message)) from ex
+            except WebsocketError as ex:
+                # Maxed out the number of retries.
+                if current_retry >= retry_max:
+                    raise ex
+                # Job id not found. No point of retrying.
+                if connection_closed_ex_code == 4003:
+                    raise ex
+
+                current_retry = current_retry + 1
+                # Retry: Continue next iteration after finally is executed.
+                continue
```

Comment:
Interestingly, a `finally` clause is executed "on the way" if it encounters certain statements, such as a `continue` statement within a `try` or `except` clause. Essentially, this means that the `finally` clause will be executed right after the `continue` statement and right before the next iteration of the loop it lives within. In this case, the websocket connection will be closed before attempting to grab another websocket connection and retrying.